File: leetcode/medium/1329. Sort the Matrix Diagonally/solution.cc
#include <iostream>
#include <vector>
#include <unordered_map>
#include <deque>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int rows = mat.size(), cols = mat[0].size();
        unordered_map<int, deque<int>> umap;

        for (int i = 0; i < rows; i += 1) {
            for (int j = 0; j < cols; j += 1) {
                umap[i-j].push_back(mat[i][j]);
            }
        }

        for (auto it: umap) {
            sort(umap[it.first].begin(), umap[it.first].end());
        }

        for (int i = 0; i < rows; i += 1) {
            for (int j = 0; j < cols; j += 1) {
                mat[i][j] = umap[i-j].front();
                umap[i-j].pop_front();
            }
        }

        return mat;
    }
};

int main() {
    Solution solution;
    vector<vector<int>> mat = {
        {3, 3, 1, 1},
        {2, 2, 1, 2},
        {1, 1, 1, 2}
    };

    vector<vector<int>> res = solution.diagonalSort(mat);

    for (int i = 0; i < res.size(); i += 1) {
        for (int j = 0; j < res[0].size(); j += 1) {
            cout << res[i][j] << " ";
        }
        cout << endl;
    }
    return 0;
}

File: leetcode/medium/1329. Sort the Matrix Diagonally/solution.js
// https://leetcode.com/problems/sort-the-matrix-diagonally/
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
// function diagonalSort(mat) {
//     const rows = mat.length;
//     const cols = mat[0].length;
//     const res = Array.from(Array(rows), () => Array(cols));

//     for (let i = 0; i < rows; i += 1) {
//         let temp = [];
//         for (let row = i, col = 0; row < rows && col < cols; row += 1, col += 1) {
//             temp.push(mat[row][col]);
//         }
//         temp.sort((a, b) => a - b);

//         for (let row = i, col = 0, j = 0; row < rows && col < cols; row += 1, col += 1, j += 1) {
//             res[row][col] = temp[j];
//         }
//     }

//     for (let i = 1; i < cols; i += 1) {
//         let temp = [];
//         for (let row = 0, col = i; row < rows && col < cols; row += 1, col += 1) {
//             temp.push(mat[row][col]);
//         }
//         temp.sort((a, b) => a - b);
//         for (let row = 0, col = i, j = 0; row < rows && col < cols; row += 1, col += 1, j += 1) {
//             res[row][col] = temp[j];
//         }
//     }

//     console.log(res)
// };

function diagonalSort(mat) {
    const map = new Map();

    for (let row = 0; row < mat.length; row += 1) {
        for (let col = 0; col < mat[0].length; col += 1) {
            const key = row - col;
            map.has(key) ? map.set(key, [...map.get(key), mat[row][col]]) : map.set(key, [mat[row][col]]);
        }
    }

    for (const [key, val] of map) {
        val.sort((a, b) => a - b).reverse();
    }

    for (let row = 0; row < mat.length; row += 1) {
        for (let col = 0; col < mat[0].length; col += 1) {
            const key = row - col;
            mat[row][col] = map.get(key).pop();
        }
    }

    console.log(mat)
    return mat;
}

diagonalSort([[3, 3, 1, 1], [2, 2, 1, 2], [1, 1, 1, 2]])